### DISCUSSION TITLE
What is rate-limiting configuration on getoutline.com?

### DISCUSSION THREAD

**Discussion by jledentu** [outline/outline#6458] [General] [answered]
Hi,

Our organization uses the Outline API to retrieve some pages and generate static HTML content. It was working perfectly until today, when we met 429 (`Rate limit exceeded`) errors.

I'd like to know how Rate Limiting is configured on getoutline.com. I was not able to find the information in the API documentation. I'm curious because I don't think we call the API so much.

Here are the headers I received:

```
Retry-After: 59.693
RateLimit-Limit: 5
RateLimit-Remaining: 0
RateLimit-Reset: Wed Jan 31 2024 09:42:41 GMT+0000 (Coordinated Universal Time)
```

So the limit would be 5 calls by minute? Seems very low ??

Thanks!

**Comment by tommoor** [comment]
The rate limit depends on the API endpoint, which endpoint are you hitting limits on?

**Reply by jledentu** [reply]
```
POST /documents.export
```

https://www.getoutline.com/developers#tag/Documents/paths/~1documents.export/post

**Reply by tommoor** [reply]
https://github.com/outline/outline/blob/main/server/routes/api/documents/documents.ts#L519-L520

The limit is 5/minute as this endpoint is computationally expensive (PDF generation is performed through this endpoint). In hindsight it might be nice to offer a higher limit to HTML exports, but that would require some refactoring to only apply to the rate limit depending on input parameters.

**Reply by jledentu** [reply]
Thanks for the explanation, even if I don't understand why I didn't reach the rate limit until now. Our tooling is suddenly broken. So is there any way to retrieve the markdown content of a set (I mean more than 5) of documents via API ?

**Reply by tommoor** [reply, marked as answer]
You can either:

- Look at `collections.export` for the equivalent method on an entire collection, this is async though – so more complex to get the result.
- Update your code to respect the rate limit – there is a `Retry-After` header which gives you the number of ms you need to wait until next request (easiest).
- We can also adjust the rate limit here in the API to be higher for non-PDF exports, but it might take a day or two given the aforementioned refactor that would be required.

**Reply by jledentu** [reply]
Thanks @tommoor, a higher limit would be nice, for now I will try the options you exposed. Thanks again for your time.

**Reply by tommoor** [reply]
Limit is upped to 25/minute now, you should still handle the rate limit headers though 😉

**Reply by jledentu** [reply]
@tommoor Thank you so much! I now handle the rate limit headers. However, the new limit is less strict so the processing will stay fast.
